modules/indexing/src/main/scala/ch/epfl/bluebrain/nexus/kg/indexing/organizations/OrganizationEsIndexer.scala

```diff
@@ -0,0 +1,129 @@
+package ch.epfl.bluebrain.nexus.kg.indexing.organizations
+
+import java.util.regex.Pattern.quote
+
+import akka.http.scaladsl.model.StatusCodes
+import cats.MonadError
+import cats.instances.string._
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.show._
+import ch.epfl.bluebrain.nexus.commons.es.client.ElasticClient
+import ch.epfl.bluebrain.nexus.commons.es.client.ElasticFailure.ElasticClientError
+import ch.epfl.bluebrain.nexus.commons.http.JsonOps._
+import ch.epfl.bluebrain.nexus.commons.iam.acls.Meta
+import ch.epfl.bluebrain.nexus.commons.test.Resources
+import ch.epfl.bluebrain.nexus.kg.core.contexts.Contexts
+import ch.epfl.bluebrain.nexus.kg.core.organizations.OrgEvent.{OrgCreated, OrgDeprecated, OrgUpdated}
+import ch.epfl.bluebrain.nexus.kg.core.organizations.{OrgEvent, OrgId}
+import ch.epfl.bluebrain.nexus.kg.indexing.ElasticIds._
+import ch.epfl.bluebrain.nexus.kg.indexing.IndexingVocab.JsonLDKeys._
+import ch.epfl.bluebrain.nexus.kg.indexing.IndexingVocab.PrefixMapping
+import ch.epfl.bluebrain.nexus.kg.indexing.IndexingVocab.PrefixMapping._
+import ch.epfl.bluebrain.nexus.kg.indexing.Qualifier._
+import ch.epfl.bluebrain.nexus.kg.indexing.jsonld.IndexJsonLdSupport._
+import ch.epfl.bluebrain.nexus.kg.indexing.{ConfiguredQualifier, Qualifier}
+import io.circe.Json
+import journal.Logger
+
+/**
+  * Organization incremental indexing logic that pushes data into an ElasticSearch indexer.
+  *
+  * @param client   the ElasticSearch client to use for communicating with the ElasticSearch indexer
+  * @param contexts the context operation bundle
+  * @param settings the indexing settings
+  * @tparam F the monadic effect type
+  */
+class OrganizationEsIndexer[F[_]](client: ElasticClient[F],
+                                  contexts: Contexts[F],
+                                  settings: OrganizationEsIndexingSettings)(implicit F: MonadError[F, Throwable])
+    extends Resources {
+
+  private val log                                                      = Logger[this.type]
+  private val OrganizationEsIndexingSettings(prefix, t, base, baseVoc) = settings
+
+  private implicit val orgIdQualifier: ConfiguredQualifier[OrgId]   = Qualifier.configured[OrgId](base)
+  private implicit val stringQualifier: ConfiguredQualifier[String] = Qualifier.configured[String](baseVoc)
+
+  private val revKey               = "rev".qualifyAsString
+  private val deprecatedKey        = "deprecated".qualifyAsString
+  private val orgName              = "name".qualifyAsString
+  private lazy val indexJson: Json = jsonContentOf("/es-index.json", Map(quote("{{type}}") -> t))
+
+  private def createIndexIfNotExist(id: OrgId): F[Unit] = {
```

Comment:
this needs to be cached as we can't afford checking the index for every event; an alternative would be to attempt to create a document and recover with index creation.

Reply:
Yes I agree. I was gonna do it with a `Map`.